packages/test/src/test-worker-lifecycle.ts

```diff
@@ -54,6 +75,59 @@ if (RUN_INTEGRATION_TESTS) {
       }
     );
   });
+
+  test.serial('Threaded VM gracely stops and fails on ERR_WORKER_OUT_OF_MEMORY', async (t) => {
+    // We internally use a timeout of 10s to catch a possible case where test would
+    // be non-conclusive. We need the test timeout to be longer than that.
+    t.timeout(30_000);
+
+    const taskQueue = t.title.replace(/ /g, '_');
+    const client = new Client();
+    const worker = await Worker.create({
+      ...defaultOptions,
+      taskQueue,
+    });
+
+    // This workflow will allocate large block of memory, hopefully causing a ERR_WORKER_OUT_OF_MEMORY.
+    // Note that due to the way Node/V8 optimize byte code, its possible that this may trigger
+    // other type of errors, including some that can't be intercepted cleanly.
+    client.workflow
```

Comment:
Hmm... is there some kind of externally generated exception that would work that isn't this? Forcing an OOM is a bit annoying and the comments are already referencing some unreliability. Maybe some kind of forced segfault or something?

Reply:
Or alternatively could we start a VM with a much smaller max allowed heap size?

Reply:
We also have a test for Uncaught Promise Rejection elsewhere.

But I absolutely want to cover the ERR_WORKER_OUT_OF_MEMORY case, even if I can't guarantee reproduction 100% of the time. It is an error that our users see in practice, some more than once a day, and until now, might leaving the Worker in a kind of zombie state.

> Or alternatively could we start a VM with a much smaller max allowed heap size?

I did at some point, and was considering exposing new worker settings for these limits. But surprisingly, that actually turned out to be no more reliable than the present test, and I decided not to expose those settings, or at least not at this time, as figuring out appropriate values turned out to be significantly harder than what it would seem.

And by the way, with the current settings, the ERR_WORKER_OUT_OF_MEMORY error happens > 95% of the time.

Reply:
Works for me